File: goldencheck/engine/scheduler.py
"""Scheduled scans — run goldencheck on a cron schedule."""
from __future__ import annotations

import logging
import time
from datetime import datetime
from pathlib import Path

logger = logging.getLogger(__name__)

__all__ = ["run_schedule"]

# Simple cron-like intervals
_INTERVALS = {
    "hourly": 3600,
    "daily": 86400,
    "weekly": 604800,
    "5min": 300,
    "15min": 900,
    "30min": 1800,
}


def run_schedule(
    files: list[Path],
    interval: str = "daily",
    domain: str | None = None,
    webhook: str | None = None,
    notify_on: str = "grade-drop",
    json_output: bool = False,
) -> None:
    """Run scheduled scans at a fixed interval.

    Args:
        files: Files to scan
        interval: "hourly", "daily", "weekly", "5min", "15min", "30min", or seconds as int
        domain: Domain pack
        webhook: Webhook URL for notifications
        notify_on: Notification trigger
        json_output: Output as JSON
    """
    import signal
    import sys
    from goldencheck.engine.scanner import scan_file
    from goldencheck.engine.confidence import apply_confidence_downgrade
    from goldencheck.engine.history import record_scan
    from goldencheck.models.finding import Severity
    from goldencheck.reporters.json_reporter import report_json as report_json_fn

    if isinstance(interval, str) and interval in _INTERVALS:
        interval_secs = _INTERVALS[interval]
    else:
        try:
            interval_secs = int(interval)
        except (ValueError, TypeError):
            raise ValueError(f"Invalid interval: {interval}. Use: {', '.join(_INTERVALS.keys())} or seconds.")

    shutdown = False

    def _handle_signal(signum, frame):
        nonlocal shutdown
        shutdown = True

    signal.signal(signal.SIGINT, _handle_signal)
    signal.signal(signal.SIGTERM, _handle_signal)

    def _ts():
        return datetime.now().strftime("[%H:%M:%S]")

    print(f"{_ts()} GoldenCheck scheduler started — scanning {len(files)} file(s) every {interval}")

    run_count = 0
    while not shutdown:
        run_count += 1
        print(f"{_ts()} Run #{run_count}")

        for file in files:
            if shutdown:
                break
            try:
                findings, profile = scan_file(file, domain=domain)
                findings = apply_confidence_downgrade(findings, llm_boost=False)

                errors = sum(1 for f in findings if f.severity == Severity.ERROR)
                warnings = sum(1 for f in findings if f.severity == Severity.WARNING)
                grade, _ = profile.health_score()

                record_scan(file, profile, findings)

                if json_output:
                    report_json_fn(findings, profile, sys.stdout)
                else:
                    print(f"  {file.name}: {errors} errors, {warnings} warnings ({grade})")

                # Webhook
                if webhook:
                    from goldencheck.engine.history import get_previous_scan
                    from goldencheck.engine.notifier import should_notify, send_webhook

                    by_col: dict[str, dict[str, int]] = {}
                    for f in findings:
                        if f.severity >= Severity.WARNING:
                            by_col.setdefault(f.column, {"errors": 0, "warnings": 0})
                            key = "errors" if f.severity == Severity.ERROR else "warnings"
                            by_col[f.column][key] = by_col[f.column].get(key, 0) + 1
                    g, s = profile.health_score(findings_by_column=by_col)

                    prev = get_previous_scan(file)
                    if should_notify(g, findings, prev, notify_on):
                        send_webhook(webhook, str(file), g, s, findings, notify_on,
                                     previous_grade=prev.grade if prev else None)

            except Exception as e:
                logger.warning("Failed to scan %s: %s", file, e)
                print(f"  {file.name}: ERROR — {e}")

        # Wait for next run
        for _ in range(interval_secs):
            if shutdown:
                break
            time.sleep(1)

    print(f"\n{_ts()} Scheduler stopped after {run_count} run(s).")

File: packages/goldencheck-js/src/core/engine/scheduler.ts
/**
 * Scheduled scans — run goldencheck on a repeating interval.
 * Port of goldencheck/engine/scheduler.py.
 * Node-only (uses node:fs via reader, signal handling).
 */

import type { Finding } from "../types.js";
import { Severity, healthScore } from "../types.js";
import { scanData } from "./scanner.js";
import { applyConfidenceDowngrade } from "./confidence.js";
import { recordScan, getPreviousScan } from "./history.js";
import { shouldNotify, sendWebhook } from "./notifier.js";
import { reportJson } from "../reporters/json.js";

export interface ScheduleOptions {
  files: readonly string[];
  interval: string; // "hourly", "daily", "weekly", "5min", "15min", "30min", or seconds as string
  domain?: string;
  webhook?: string;
  notifyOn?: string;
  jsonOutput?: boolean;
}

// Named interval mappings (seconds)
const INTERVALS: Record<string, number> = {
  hourly: 3600,
  daily: 86400,
  weekly: 604800,
  "5min": 300,
  "15min": 900,
  "30min": 1800,
};

function parseInterval(interval: string): number {
  if (interval in INTERVALS) {
    return INTERVALS[interval]!;
  }
  const parsed = parseInt(interval, 10);
  if (isNaN(parsed) || parsed <= 0) {
    throw new Error(
      `Invalid interval: ${interval}. Use: ${Object.keys(INTERVALS).join(", ")} or seconds.`,
    );
  }
  return parsed;
}

function timestamp(): string {
  const now = new Date();
  return `[${now.toTimeString().slice(0, 8)}]`;
}

/**
 * Sleep for the given number of seconds, checking the abort signal every second.
 * Resolves to true if sleep completed, false if aborted.
 */
function sleepWithAbort(seconds: number, signal: AbortSignal): Promise<boolean> {
  return new Promise<boolean>((resolve) => {
    let elapsed = 0;
    const tick = () => {
      if (signal.aborted || elapsed >= seconds) {
        resolve(!signal.aborted);
        return;
      }
      elapsed += 1;
      setTimeout(tick, 1000);
    };
    tick();
  });
}

/**
 * Run scheduled scans at a fixed interval.
 * Scans each file, records history, prints results, fires webhooks.
 * Gracefully shuts down on SIGINT/SIGTERM.
 */
export async function runSchedule(options: ScheduleOptions): Promise<void> {
  const intervalSecs = parseInterval(options.interval);
  const notifyOn = options.notifyOn ?? "grade-drop";

  // AbortController for graceful shutdown
  const ac = new AbortController();

  const onSignal = () => {
    ac.abort();
  };
  process.on("SIGINT", onSignal);
  process.on("SIGTERM", onSignal);

  console.log(
    `${timestamp()} GoldenCheck scheduler started — scanning ${options.files.length} file(s) every ${options.interval}`,
  );

  // Dynamic import of reader (Node-only)
  let readFile: (path: string) => import("../data.js").TabularData;
  try {
    const reader = await import("../../node/reader.js");
    readFile = reader.readFile;
  } catch {
    throw new Error(
      "Scheduler requires the Node reader. Import from 'goldencheck/node' or ensure node/reader.js is available.",
    );
  }

  let runCount = 0;
  while (!ac.signal.aborted) {
    runCount += 1;
    console.log(`${timestamp()} Run #${runCount}`);

    for (const file of options.files) {
      if (ac.signal.aborted) break;

      try {
        const data = readFile(file);
        const result = scanData(data, { domain: options.domain ?? undefined });
        let findings = applyConfidenceDowngrade(result.findings, false);

        // Fix filePath in profile
        const profile = { ...result.profile, filePath: file };

        const errors = findings.filter((f) => f.severity === Severity.ERROR).length;
        const warnings = findings.filter((f) => f.severity === Severity.WARNING).length;

        // Compute health score for display and webhook
        const byCol: Record<string, { errors: number; warnings: number }> = {};
        for (const f of findings) {
          if (f.severity >= Severity.WARNING) {
            if (!byCol[f.column]) {
              byCol[f.column] = { errors: 0, warnings: 0 };
            }
            if (f.severity === Severity.ERROR) {
              byCol[f.column]!.errors += 1;
            } else {
              byCol[f.column]!.warnings += 1;
            }
          }
        }
        const hs = healthScore(byCol);

        // Record to history
        recordScan(file, profile, findings);

        // Output
        if (options.jsonOutput) {
          console.log(reportJson(findings, profile));
        } else {
          const fileName = file.split("/").pop() ?? file;
          console.log(`  ${fileName}: ${errors} errors, ${warnings} warnings (${hs.grade})`);
        }

        // Webhook notification
        if (options.webhook) {
          const prev = getPreviousScan(file);
          if (shouldNotify(hs.grade, findings, prev, notifyOn)) {
            await sendWebhook(
              options.webhook,
              file,
              hs.grade,
              hs.points,
              findings,
              notifyOn,
              prev?.grade,
            );
          }
        }
      } catch (e) {
        const fileName = file.split("/").pop() ?? file;
        console.error(`  ${fileName}: ERROR — ${e}`);
      }
    }

    // Wait for next run (check abort every second)
    const completed = await sleepWithAbort(intervalSecs, ac.signal);
    if (!completed) break;
  }

  // Cleanup signal handlers
  process.removeListener("SIGINT", onSignal);
  process.removeListener("SIGTERM", onSignal);

  console.log(`\n${timestamp()} Scheduler stopped after ${runCount} run(s).`);
}
